Regina Tamayo León

23 abril 2026

## **Árbol de regresión para predicciones**

In [28]:
import pandas as pd

datos = [
    {"genero": "H", "edad": 21, "estatura": 1.73},
    {"genero": "M", "edad": 19, "estatura": 1.62},
    {"genero": "H", "edad": 21, "estatura": 1.73},
    {"genero": "H", "edad": 19, "estatura": 1.69},
    {"genero": "H", "edad": 20, "estatura": 1.75},
    {"genero": "H", "edad": 21, "estatura": 1.80},
    {"genero": "H", "edad": 22, "estatura": 1.70},
    {"genero": "H", "edad": 21, "estatura": 1.75},
    {"genero": "H", "edad": 19, "estatura": 1.66},
    {"genero": "M", "edad": 21, "estatura": 1.64},
    {"genero": "M", "edad": 19, "estatura": 1.62},
    {"genero": "H", "edad": 21, "estatura": 1.85},
    {"genero": "H", "edad": 20, "estatura": 1.75},
    {"genero": "H", "edad": 20, "estatura": 1.66},
    {"genero": "H", "edad": 19, "estatura": 1.70},
    {"genero": "H", "edad": 19, "estatura": 1.64}
]

df = pd.DataFrame(datos)

df_h_menor_21 = df[(df["genero"] == "H") & (df["edad"] < 21)]
print("Hombres menores a 21 años")
print(f"- Estatura promedio: {df_h_menor_21['estatura'].mean():.2f} m")

df_h_mayor_21 = df[(df["genero"] == "H") & (df["edad"] >= 21)]
print("Hombres de 21 años o más")
print(f"- Estatura promedio: {df_h_mayor_21['estatura'].mean():.2f} m")

df_m_19 = df[(df["genero"] == "M") & (df["edad"] == 19)]
print("Mujeres 19")
print(f"- Estatura promedio: {df_m_19['estatura'].mean():.2f} m")

df_m_21 = df[(df["genero"] == "M") & (df["edad"] == 21)]
print("Mujeres 21")
print(f"- Estatura promedio: {df_m_21['estatura'].mean():.2f} m")

Hombres menores a 21 años
- Estatura promedio: 1.69 m
Hombres de 21 años o más
- Estatura promedio: 1.76 m
Mujeres 19
- Estatura promedio: 1.62 m
Mujeres 21
- Estatura promedio: 1.64 m


### **Actividad**

In [31]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

df = pd.read_csv('Advertising.csv')
df.head()

,Unnamed: 0,TV,radio,newspaper,sales
0,1,230.1,37.8,69.2,22.1
1,2,44.5,39.3,45.1,10.4
2,3,17.2,45.9,69.3,9.3
3,4,151.5,41.3,58.5,18.5
4,5,180.8,10.8,58.4,12.9


**Regresión lineal simple**

In [33]:
X = df[['TV', 'radio', 'newspaper']]
y = df['sales']

modelo = LinearRegression()
modelo.fit(X, y)

m = modelo.coef_[0]      
b = modelo.intercept_    
print(f"Ecuación del modelo: Ventas = {m:.4f} * TV + {b:.4f}")

Ecuación del modelo: Ventas = 0.0458 * TV + 2.9389


**Cálculo R2** 

In [34]:
r2_lineal = modelo.score(X, y)

print(f"El R2 de la Regresión Lineal Simple es: {r2_lineal:.4f}")

El R2 de la Regresión Lineal Simple es: 0.8972


**Árbol de Decisión para regresión**

In [36]:
modelo_arbol = DecisionTreeRegressor(random_state=42)

modelo_arbol.fit(X, y)

DecisionTreeRegressor(random_state=42)

In [37]:
r2_arbol = modelo_arbol.score(X, y)

print(f"El R2 del Árbol de Decisión es: {r2_arbol:.4f}")

El R2 del Árbol de Decisión es: 1.0000


**Train-Test Split 50/50**

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.50, random_state=42)

modelo_lin_split = LinearRegression()
modelo_lin_split.fit(X_train, y_train) 
r2_lin_split = modelo_lin_split.score(X_test, y_test) 

modelo_arbol_split = DecisionTreeRegressor(random_state=42)
modelo_arbol_split.fit(X_train, y_train) 
r2_arbol_split = modelo_arbol_split.score(X_test, y_test) 

print("Resultados Train/Test Split 50/50")
print(f"R2 Regresión Lineal en Prueba: {r2_lin_split:.4f}")
print(f"R2 Árbol de Decisión en Prueba: {r2_arbol_split:.4f}")

Resultados Train/Test Split 50/50
R2 Regresión Lineal en Prueba: 0.8721
R2 Árbol de Decisión en Prueba: 0.9345


Incluso frente a datos desconocidos (Test), el Árbol de Decisión demostró ser superior para este caso específico, obteniendo un R² de 0.9345 frente al 0.8721 de la Regresión Lineal. Esto significa que el árbol logró descifrar un 93.4% del comportamiento de las ventas, capturando interacciones más complejas entre la inversión en TV, radio y periódico que una simple línea recta no pudo detectar.